src/phoenix/trace/llama_index.py

```diff
@@ -0,0 +1,172 @@
+        self._tracer = Tracer(on_append=callback)
+        self._event_id_to_event_data: Dict[str, CBEventData] = defaultdict(lambda: CBEventData())
+
+    def on_event_start(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> str:
+        event_id = event_id or str(uuid4())
+        self._event_id_to_event_data[event_id]["name"] = event_type.value
+        self._event_id_to_event_data[event_id]["start_event"] = CBEvent(
+            event_type=event_type,
+            payload=payload,
+            id_=event_id,
+        )
+        return event_id
+
+    def on_event_end(
+        self,
+        event_type: CBEventType,
+        payload: Optional[Dict[str, Any]] = None,
+        event_id: str = "",
+        **kwargs: Any,
+    ) -> None:
+        event_id = event_id or str(uuid4())
+        self._event_id_to_event_data[event_id].setdefault("name", event_type.value)
+        self._event_id_to_event_data[event_id]["end_event"] = CBEvent(
+            event_type=event_type,
+            payload=payload,
+            id_=event_id,
+        )
+
+    def start_trace(self, trace_id: Optional[str] = None) -> None:
+        ...
+
+    def end_trace(
+        self,
+        trace_id: Optional[str] = None,
+        trace_map: Optional[Dict[str, List[str]]] = None,
+    ) -> None:
+        _add_to_tracer(
+            event_id_to_event_data=self._event_id_to_event_data,
+            child_to_parent_trace_id=dict(
+                _generate_child_to_parent_trace_id_pairs(trace_map or {})
+            ),
+            tracer=self._tracer,
+        )
+        self._event_id_to_event_data = defaultdict(lambda: CBEventData())
+
+
+def _add_to_tracer(
+    event_id_to_event_data: Dict[str, CBEventData],
+    child_to_parent_trace_id: Dict[str, str],
+    tracer: Tracer,
+) -> None:
+    """Adds event data to the tracer, where it is converted to a span and stored in a buffer.
+
+    Args:
+        event_id_to_event_data (Dict[str, CBEventData]): A map of event IDs to event data.
+
+        child_to_parent_trace_id (Dict[str, str]): A map of child trace IDs to parent trace IDs.
+
+        tracer (Tracer): The tracer that stores spans.
+    """
+    for event_id, event_data in event_id_to_event_data.items():
+        parent_event_id = child_to_parent_trace_id.get(event_id)
+        start_event = event_data.get("start_event")
+        end_event = event_data.get("end_event")
+        name = event_data.get("name", "")
+        if start_event:
+            span_kind = _get_span_kind(start_event.event_type)
+        elif end_event:
+            span_kind = _get_span_kind(end_event.event_type)
+        else:
+            span_kind = SpanKind.CHAIN
+        tracer.create_span(
+            name=name,
+            span_kind=span_kind,
+            start_time=start_event.time if start_event else None,
+            end_time=end_event.time if end_event else None,
```

Comment:
we may not want these to be optional. they should be filled one way or another

Reply:
Good catch, not sure how this is passing type checks. In the case where the start event or end event is not logged, I'd like to record that information somehow, though.

Reply:
true, but for mvp, we can assume it works for now and deal with the outliers later, since we don't know if they actually exist (or how)

Reply:
addressed in [0c39152](https://github.com/Arize-ai/phoenix/pull/1058/commits/0c391523f8ad2198d8281e3ce35b456e6960f913)